Added event System!

ridhwandll opened the pull request:
# Added basic Event System
- The events are short time tweaks to some normal configurations or other added functionality to increase the *FUN* 
- Admins can select who can launch or end an event by `!give-new-event-access @Roles`!
- Currently there are two events, `HighRiskHighReward` and `LowRiskLowReward`, they change the copy of the real `stealSuccessChance`!
- Time of an event can be set by the third argument of the `!lanuch` command, events will be automatically ended when the time ends! The `@Roles` who have permission to launch and end events can end the events forcibly by `!endevent`
- The event system has its separate module!
# Update
- New event RaffleDraw!
- The admins and the selected @roles can launch this event by `!launch RaffleDraw <time>`
- The users can buy tickets by `!buyticket <amount>` when the RaffleDraw event starts!
- As always, the admins and the selected @roles can end this event forcibly by !endevent.
- When the event ends, the bot will announce the lucky ticket, and the user have to get the rewards by `!getreward`
- If the user fails to get the reward within 24 hours the reward will be gone, and the admin and the selected @roles can launch RaffleDraw again!
- Currently the price of the RaffleDraw is 500 CC, It can be updated later.
## Fix
- A fix and some improvements for the `!roll` command is added in this pull request. (Finally it won't crash anymore). 
- Some improvement in the bots response in the `!bribe` command.
# How to use this event system?
- To start an event write `!launch <event-code-or-name> <expiry-time-in-hours>`. There are currently 3 events.
- HighRiskHighReward(code to launch: `HRHR`)
   - The steal chance is **DECREASED** for the time being! If the thief succeeds in the steal, he will get a **BONUS** CC which is proportional to the amount he has stolen.
- LowRiskLowReward (code to launch: `LRLR`)
   - The steal chance is **INCREASED** for the time being! If the thief succeeds in the steal, he will have to give a **PENALTY** which is proportional to the amount he has stolen.
- RaffleDraw (code to launch:[No code] just type `RaffleDraw` to launch)
   - The most **EXCITING** event in my opinion!<br/>
   The users can buy tickets with the command `!buyticket <amount of tickets>` (This command will be moved to the shop, when the shop comes!) Each ticket costs `50 CC` initially, this can be changed by the admin command `!setticketprice <amount>`(another thing that will be gone when the shop comes). The users can buy `20` tickets at max initially, can be set buy using the admin command `!setusermaxticket <amount>`. *Each ticket has an unique id*
   - **So why buy this tickets?**
      - The result of the draw will be announced automatically when the event ends, here the result means the ticket id.
      - The user who has the lucky ticket can use the command `!getreward` to get their rewards. Currently the reward is `500 CC` that can be updated later when we get the shop(maybe we can giveaway some shop items?)
   - **NOTE: Admins cannot launch another RaffleDraw event after the end of a RaffleDraw until the lucky user claims their reward. The lucky user have 24 hours to claim their reward, if they don't collect it within 24 hours, rewards will be gone and the admins can start the RaffleDraw event again. *The admins can start this event immidiatley after the lucky user have claimed their reward.* The other events such as the HRHR or LRLR can be launched. They have no relation with the RaffleDraw reward is claimed or not.**
## `!endevent` command
- The event will be ended forcibly! If the RaffleDraw was running then it will announce the lucky ticket immidiatedly. **Can only be used by the `@roles` who have permission to launch or endevents!**
### Event Access management(the roles to launch or end events)
- Use the command `!give-new-event-access @Role` to give access to a certain role(They can launch or end events).
- Use command `!take-new-event-access @Role` to take access from a Role(they can no longer launch or end events).
- **NOTE: These are only the admin commands, only the admins can use this.**

ridhwandll committed:
Added event System!

ridhwandll committed:
Minor improvements to the event system.

Added more juice to bots response regarding events!
Some fixes to the event system

Gaztin commented on src/modules/CurrencyModule.cpp:
```diff
@@ -914,21 +915,113 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		EventModule* eventModule = static_cast<EventModule*>(UmikoBot::Instance().GetModuleByName("event"));
+		auto& eventConfig = eventModule->getEventServerData(channel.guildId());
+		
+		if (eventConfig.eventHighRiskHighRewardRunning)
```
When is this event enabled exactly?

ridhwandll replied:
# HighRiskHighReward
**The steal chance is decreased, but if the users succeed on stealing then they will get BONUS CC**
- The BONUS is proportionate to the amount of the CC that a user steals.
- It enables when you give the command `!launch HRHR <amount-of-time-in-hours>`, Launching all the event is same, you just have to remember the event name 🙂

Gaztin replied:
*When* is it enabled, though?

Gaztin replied:
I can't see where you change this boolean.

Gaztin replied:
Never mind, the diff was closed

ridhwandll committed:
Started working on RaffleDraw event

ridhwandll committed:
Added RaffleDraw event!

Gaztin reviewed (changes requested):
I'm sorry, but with the raffle commands, it's a tad too much to review. Please put those commits in another branch revert those commits from this PR.

ridhwandll commented:
@Gaztin I AM SORRY, because I didn't make them as different, I will follow this from next time!

ridhwandll committed:
Small fixes in the RaffleDraw event 


Added some juice to the bot's response.


Some changes.

ridhwandll committed:
Now the admins and the selected roles can launch events for more than 1day

ridhwandll committed:
A major fix in the RaffleDraw event.

ridhwandll committed:
Build fix for Linux (event: RaffleDraw)


Hopefully Linux build error is fixed.


Changed the way of random TicketID generation for fixing build error.

ridhwandll committed:
Some polishing to the bots response regarding the RaffleDraw event

Added the limit of the tickets, that users can buy.
The set user max ticket is turned into an admin command.

NotJustSid commented on res/commands.json:
```diff
@@ -270,16 +270,69 @@
+		"admin": true
+	},
+	"EVENT_SET_USER_MAX_TICKET": {
+		"brief": "Sets the max ticket amount a user can buy.",
```
I might be being picky, but I'd say change this to
```
Sets the maximum tickets a user can buy.
```
Clean and concise...

ridhwandll replied:
Perfectly Done! 🙂

NotJustSid commented on src/core/Utility.h:
```diff
@@ -81,6 +81,11 @@ namespace utility
 			}
 			
+			constexpr auto TICKETS = ":tickets:";
+			constexpr auto REGIONAL_INDICATOR_E = ":regional_indicator_e:";
+			constexpr auto REGIONAL_INDICATOR_V = ":regional_indicator_v:";
+			constexpr auto REGIONAL_INDICATOR_N = ":regional_indicator_n:";
+			constexpr auto REGIONAL_INDICATOR_T = ":regional_indicator_t:";
```
I'd say add all of the `REGIONAL_INDICATORS` instead of just these.. Those are already under `reacts::` so you need to just copy and paste them.

ridhwandll replied:
Okay I added all of them, (a - z). 🙂

NotJustSid commented on src/modules/CurrencyModule.h:
```diff
@@ -4,10 +4,15 @@
+//Used for the events (Needed in the EventModule.cpp and CurrencyModule.cpp)
+#define highRiskRewardBonus 50 // In percentage
+#define lowRiskRewardPenalty 40 //In percentage
+
```
These defines would be better suited under `EventModule.h`

NotJustSid commented on src/modules/CurrencyModule.h:
```diff
@@ -110,6 +115,9 @@ class CurrencyModule : public Module
 		int bribeLeastAmount { 20 };
 		int bribeSuccessChance { 68 };
+
+		int lowRiskRewardStealSuccessChance{ 50 };
+		int highRiskRewardStealSuccessChance{ 30 };
```
Again, semantically speaking, these variables shouldn't be here.

ridhwandll replied:
I think they should be in the `CurrencyModule.cpp`. No other events are using them and they only effect the CurrencySystem. I removed the defines to the `EventModule.h` because they are needed by both of the modules. But in this case I don't think its needed to move them. 🙂

NotJustSid commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,72 @@
+		bool eventLowRiskLowRewardRunning{ false };
+		bool claimedReward = true;
+		bool eventRaffleDrawRunning{ false };
+		unsigned int currentTicketIndex{ 0 };
```
This variable stands out. What is this used for?

NotJustSid replied:
Edit: Okay I see how you use it, but it could be named better. You could go with something like `numticketsDispensed` or `numticketsBought` to signify that it's the total number of tickets that have been bought.

NotJustSid commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,72 @@
+		QTimer* raffleDrawRewardClaimTimer{ nullptr };
+		bool isEventRunning{ false };
+		bool eventHighRiskHighRewardRunning{ false };
+		bool eventLowRiskLowRewardRunning{ false };
```
You're not required to prefix the variables with "event". A better name would be `isHighRiskHighRewardRunning`. The "is" also clarifies that this is indeed a boolean.

NotJustSid commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,72 @@
+	{
+		QTimer* eventTimer{ nullptr };
+		QTimer* raffleDrawRewardClaimTimer{ nullptr };
+		bool isEventRunning{ false };
```
Isn't this redundant given that you have other variables to know if an event is running and what type of event it is?

ridhwandll replied:
Nope, its needed because it activates whenever an event is running that can be any event. It tracks the event system is in use or not. The event timer is based on it. 🙂

NotJustSid commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,72 @@
+	QMap<snowflake_t, QList<snowflake_t>> m_EventWhitelist;
+	QMap<snowflake_t, EventConfig>serverEventConfig;
+
+	snowflake_t raffleDrawGetUserIndex(snowflake_t guild, snowflake_t id)
```
Why does this return a `snowflake_t` (which the code base uses for ids) when it is supposed to return an `int` (the type of the index of elements in `QList`)?

NotJustSid commented on src/modules/FunModule.cpp:
```diff
@@ -70,45 +70,74 @@ FunModule::FunModule(UmikoBot* client) : Module("funutil", true), m_memeChannel(
 
 	RegisterCommand(Commands::FUN_ROLL, "roll", [this](Client& client, const Message& message, const Channel& channel)
 		{
+			QStringList args = message.content().split(' ');
```
Okay, why are changes to the FunModule which are totally unrelated to the event system in this PR?

ridhwandll replied:
See the PR comment, the roll command is updated 🙂 
![image](https://user-images.githubusercontent.com/69952230/99807256-720dce80-2b69-11eb-82c2-a5b6ae6c6951.png)

NotJustSid commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,72 @@
+		bool eventRaffleDrawRunning{ false };
+		unsigned int currentTicketIndex{ 0 };
+		int raffleDrawTicketPrice{ 50 };
+		int maxTicketOfUser{ 20 };
```
`maxUserTickets` would be a better name...

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+
+			Embed embed;
+			embed.setColor(15844367);
+			embed.setTitle("**HighRiskHighReward** event");
```
If I recall correctly, bold has little to no effect in embed titles.

NotJustSid replied:
I also think "HighRiskHighReward" (all together) in a title would look cluttered.

ridhwandll replied:
Then, you want `High Risk High Reward` ?

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+			Embed embed;
+			embed.setColor(15844367);
+			embed.setTitle("**LowRiskLowReward** event");
+			embed.setDescription("The steal chance is **increased** but you have to give a **PENALTY** if you succeed!\n"
```
"pay a penalty" sounds better 😄

ridhwandll replied:
Okay done 🙂

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+
+			Embed embed;
+			embed.setColor(15844367);
+			embed.setTitle("**LowRiskLowReward** event");
```
Same as above.

ridhwandll replied:
Same as above 😋

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+
+			Embed embed;
+			embed.setColor(15844367);
+			embed.setTitle("**RAFFLE DRAW** event");
```
Same as the comments about the titles. Also, let's not capitalize the name.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+			Embed embed;
+			embed.setColor(15844367);
+			embed.setTitle("**RAFFLE DRAW** event");
+			embed.setDescription("You can buy tickets and earn cool rewards if you win!\n"
```
We haven't really decided about the rewards have we? The reward is just some cash prize (which I think should be a small amount because we already have freebies).

ridhwandll replied:
Yes, I am decrementing it to `400`, we need a prize. I will add them when the shop comes 😄

ridhwandll replied:
EDIT: From the above comment of mine: *add them* means adding items from the shop.

ridhwandll replied:
The cash prize will get decreased heavily, when the shop item comes 😄

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+			embed.setColor(15844367);
+			embed.setTitle("**RAFFLE DRAW** event");
+			embed.setDescription("You can buy tickets and earn cool rewards if you win!\n"
+								"The results of the draw is announced when the event expires.\n"
```
Typo: "results" instead of "result"

ridhwandll replied:
Thanks for pointing that out 😄

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+			embed.setTitle("**RAFFLE DRAW** event");
+			embed.setDescription("You can buy tickets and earn cool rewards if you win!\n"
+								"The results of the draw is announced when the event expires.\n"
+								"Use `!buyticket <amount>` to buy ticket for participating.\n"
```
Since `buyticket` let's one buy multiple tickets, use "tickets" instead of "ticket".

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+								"The results of the draw is announced when the event expires.\n"
+								"Use `!buyticket <amount>` to buy ticket for participating.\n"
+								"Buying more tickets increases your success chance!\n"
+								"Event **ends** in `" + time + "`");
```
That's an unnecessary and awkward looking bold in between.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+	RegisterCommand(Commands::EVENT_GIVE_NEW_ACCESS, "give-new-event-access", [this](Client& client, const Message& message, const Channel& channel)
+	{
+		QStringList args = message.content().split(' ');
+		UmikoBot::VerifyAndRunAdminCmd(client, message, channel, 2, args, false, [this, &client, channel, message, args]()
```
You don't use `client` and `args` in the lambda... Why are these in the capture?

ridhwandll replied:
Oops, forgot to remove them after copying 😋 REMOVED 😄

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+	{
+		QStringList args = message.content().split(' ');
+
+		UmikoBot::VerifyAndRunAdminCmd(client, message, channel, 2, args, false, [this, &client, channel, message, args]()
```
Same as above.

ridhwandll replied:
Same as above.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+
+					if (config.isEventRunning)
+					{
+						UmikoBot::Instance().createMessage(message.channelId(), "An event is already running. Please stop that to start a new event!");
```
I would still say it would be awesome if we could have multiple events simultaneously.

ridhwandll replied:
That would be a complete mess in my opinion. People would mix up the events if we run multiple events simultaneously. The events are unique and kind of a surprise. Multiple events will make them very common thing. 🙂

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+					}
+					if (time > 576)
+					{
+						UmikoBot::Instance().createMessage(message.channelId(), "**You can't launch events with an expiry time of more than 24 days!**.");
```
Shorter message: "Events cannot last longer than 24 days!"

NotJustSid replied:
Also, are you going to implement the idea of creating a wrapper around a list of `QTimer`s to accommodate for larger intervals as we had discussed?

ridhwandll replied:
Nope, 24 days is enough for the events in my opinion.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+						UmikoBot::Instance().createMessage(message.channelId(), "**Wrong Usage of Command!**.");
+						return;
+					}
+					if (time > 576)
```
Better to write it like `24*24` to make it clear why you have this number here.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+					}
+					config.eventTimer = new QTimer;
+
+					config.eventTimer->setInterval(time * 3600000);
```
Again, how does `3600000` come up?

ridhwandll replied:
That's the time in milliseconds

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+					config.eventTimer->setInterval(time * 3600000);
+					config.eventTimer->setSingleShot(true);
+
+					if (args.at(1) != "HRHR" && args.at(1) != "LRLR" && args.at(1) != "RaffleDraw")
```
To make this extensible and to help conveniently add new events in the future, you should instead make a list of all the codes/names and then check if the argument is in there.

ridhwandll replied:
Okay perfectly done 😄

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+													QString(utility::consts::emojis::REGIONAL_INDICATOR_E) + " " + QString(utility::consts::emojis::REGIONAL_INDICATOR_N) + " " +
+															QString(utility::consts::emojis::REGIONAL_INDICATOR_T);
+
+					if (args.at(1) == "HRHR")
```
You said we could also use the event names. Where do you handle the case where `args.at(1)` is *"HighRiskHighReward"*?

ridhwandll replied:
I and FKP decided to make the big event names turn into a code. That *HighRiskHighReward* is very tedious to write. 🙂

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+						config.eventHighRiskHighRewardRunning = true;
+						UmikoBot::Instance().createMessage(message.channelId(), embed);
+					}
+					if (args.at(1) == "LRLR")
```
Same query as above.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+						config.eventLowRiskLowRewardRunning = true;
+						UmikoBot::Instance().createMessage(message.channelId(), embed);
+					}
+					if (args.at(1) == "RaffleDraw")
```
Why doesn't "RaffleDraw" have a corresponding code?

ridhwandll replied:
Because, it is not tedious to write 😄

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+							auto& config = getServerEventData(channel.guildId());
+							int timeLeft = config.raffleDrawRewardClaimTimer->remainingTime();
+							QString time = utility::StringifyMilliseconds(timeLeft);
+							UmikoBot::Instance().createMessage(message.channelId(), "**The winner of the previous RaffleDraw didn't claimed their rewards yet.**\n"
```
"hasn't claimed" instead of "didn't claimed".

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+									{
+										auto& config = getServerEventData(guildID);
+										config.claimedReward = true;
+										UmikoBot::Instance().createMessage(chan, "The owner of the price in the RaffleDraw didn't get their rewards!\n"
```
Should be "prize" instead of "price".

NotJustSid replied:
Also use "claim" instead of "get"

NotJustSid replied:
And it would be better if we don't reference the winner as the owner of the prize because they would have to claim it first to own it.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+					CurrencyModule* currencyModule = static_cast<CurrencyModule*>(UmikoBot::Instance().GetModuleByName("currency"));
+					auto& currencyConfig = currencyModule->getServerData(channel.guildId());
+
+					QString eventEmoji = QString(utility::consts::emojis::REGIONAL_INDICATOR_E) + " " + QString(utility::consts::emojis::REGIONAL_INDICATOR_V) + " " +
+											QString(utility::consts::emojis::REGIONAL_INDICATOR_E) + " " + QString(utility::consts::emojis::REGIONAL_INDICATOR_N) + " " +
+												QString(utility::consts::emojis::REGIONAL_INDICATOR_T);
```
You don't need to hype up the end of an event.

ridhwandll replied:
It needs to attract the eyes, if I remove the current title of the embed, Please give me a better title and a description. Yes, the description is needed I think because, the only title wouldn't attract the users attention, its so small.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+							config.eventRaffleDrawRunning = false;
+							config.eventTimer->stop(); // Forcefully stop the timer.
+							UmikoBot::Instance().createMessage(message.channelId(), embed);
+						}
```
There is a lot of code duplication in end-event. You could instead manually trigger the timer timeout and handle ending in the timeout callback.

ridhwandll replied:
Sorry, I didn't get your words.

ridhwandll replied:
EDIT: I understand the code duplication thing, but I didn't understand what you want me to do.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		auto& config = getServerEventData(channel.guildId());
+		if (config.isEventRunning)
+		{
+			client.createMessage(message.channelId(), "**You can't get free rewards anytime!**");
```
The rewards technically aren't free considering you have to pay for the tickets.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+					authorCurrency.setCurrency(authorCurrency.currency() + 500);
+					embed.setTitle("**Raffle Draw Reward goes to " + authorName + "!**");
+					embed.setDescription("Congratulations **" + authorName + "**!\n"
+										"You just got `500` " + currencyConfig.currencySymbol);
```
In almost all the other places the symbol is included inside the code block, so would be better if we did that here as well :)

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		}
+		else
+		{
+			client.createMessage(message.channelId(), "**You can't get free rewards anytime!**");
```
> The rewards technically aren't free considering you have to pay for the tickets.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		auto& configRD = raffleDrawGuildList[channel.guildId()][raffleDrawGetUserIndex(channel.guildId(), authorID)];
+		if (configRD.m_TicketIds.size() >= config.maxTicketOfUser)
+		{
+			client.createMessage(message.channelId(), "**You have already bought so many tickets, I can't give you more!**");
```
I think both of messages for these checks should be the same (the one below is almost similar, but you forgot a space after the period):
```
"**Don't be selfish!**\nLet others also have a go at some tickets. You can only have " + num + " tickets at max!"
```

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		}
+		if (totalFee > authorCurrency.currency())
+		{
+			client.createMessage(message.channelId(), "**Can you afford that much ticket?**");
```
Use "Can you afford that many tickets?" instead.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		authorCurrency.setCurrency(authorCurrency.currency() - totalFee);
+		QString ticketEmoji = QString(utility::consts::emojis::TICKETS);
+		QString ticketAmount = QString::number(configRD.m_TicketIds.size());
+		client.createMessage(message.channelId(), ticketEmoji + " **You own** `" + ticketAmount + "` **ticket(s) now!**\n *(Buy them more to increase your chance)*");
```
"Buy them more" sounds weird; use "buy more of them" instead.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+		UmikoBot::VerifyAndRunAdminCmd(client, message, channel, 2, args, true, [this, &client, channel, message, args]()
+			{
+				auto& config = getServerEventData(channel.guildId());
+				config.raffleDrawTicketPrice = args.at(1).toInt();
```
Most of the currency system currently uses `double`s (though we have plans to change it I guess), so one would assume that this also accepts `doubles`.

NotJustSid commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,792 @@
+			{
+				auto& config = getServerEventData(channel.guildId());
+				config.maxTicketOfUser = args.at(1).toInt();
+				client.createMessage(message.channelId(), "User max ticket set to **" + QString::number(config.maxTicketOfUser) + "**");
```
There aren't any global and user level max ticket amounts, so the "User" here isn't necessary.

NotJustSid commented on src/modules/CurrencyModule.cpp:
```diff
@@ -914,21 +915,113 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), output);
+			}
+			return;
+		}
```
Code duplication here. Most of the code for the three modes - "HRHR", "LRLR" and "Normal" is similar and could be combined. Create local variables instead, set them accordingly with the conditionals, and use them for the whole *success chance and friends* stuff.

ridhwandll replied:
Okay done

ridhwandll committed:
Some optimizations regarding the events.


Reduced code duplication in the Event module

ridhwandll committed:
Fixes after rebase


Fix
